### DISCUSSION TITLE
Custom highlighting

### DISCUSSION THREAD

**Discussion by kknishev** [sevdokimov/log-viewer#194] [Q&A] [answered]
Assist please, is it possible to use custom highlighting?
As I see - setup a list special tags in config.conf, something like this:
'[ALARM]' - color: red;
'[ATTANTION]' - color: yellow;
'[NORMAL]' - color: green;

it's important to use custom highlighting without the log format/pattern due to non-standard log structure

**Comment by sevdokimov** [comment]
'ALARM', 'ATTANTION', and 'NORMAL' are custom log levels specified in `log-viewer.log-levels` configuration property in `config.conf`, correct?
Could you attach the configuration and a sample of log line?

**Reply by kknishev** [reply]
```
logs = [
  {
    path: "**/*.log"  // All files with ".log" extension are visible by default
    format: {
      type: Log4jLogFormat
      locale: en_US // Set  custom locale (optional)
      pattern: "%date{MMM [ ]d HH:mm:ss} %thread %msg"
    }
  } ]
```


```
[example.log](https://github.com/sevdokimov/log-viewer/files/12498801/example.log)
```

**Reply by sevdokimov** [reply]
Unfortunately, it's impossible in the current version.

**Reply by kknishev** [reply]
will be great to get it in the next versions, thanks a lot

**Reply by dima-bzz** [reply]
@KKnishev Do you want to change the color of the log levels?

**Reply by kknishev** [reply]
finally yes, but I couldn't recognize the levels in my logs, levels is an optional part

**Reply by dima-bzz** [reply]
@KKnishev You can change the default colors. But your logs don't have logging levels everywhere.

**Reply by kknishev** [reply]
yes, that's the point why I can't use pattern: with %level

**Comment by kknishev** [comment]
@sevdokimov hi Sergey
I think another way "to get" highlighting for custom log levels is using several format:/pattern: in one configuration
(first one common and the next one more stronger)
```
pattern_1: "%date{MMM [ ]d HH:mm:ss} %thread %msg"
pattern_2: "%date{MMM [ ]d HH:mm:ss} %thread %level %msg"
```
assist please, is it possible in this way or something else? thanks a lot

**Reply by sevdokimov** [reply]
Using several patterns in one file is a good idea, but it's not supported now. I'll try to add the support of that today or tomorrow.

**Reply by sevdokimov** [reply]
@KKnishev Please, try 1.0.7, it supports multipattern Log4j Format. You can specify the following configuration:
```
logs = [
  // ...

  {
    path: "/home/foo/bar/my-log-file.log"
    format: {
      type: Log4jLogFormat
      patterns: [
        "%date{MMM [ ]d HH:mm:ss} %thread %msg%n",
        "%date{MMM [ ]d HH:mm:ss} %thread %level %msg%n"
      ]
     customLevels: ["ALARM", "ATTANTION", "NORMAL"]
    }
  }

 // ...
]

```

**Reply by kknishev** [reply]
@sevdokimov great, waiting for builds, please...

**Comment by kknishev** [comment]
@sevdokimov morning sergey, could you please add a 1.0.7 jar builds
regards

**Reply by sevdokimov** [reply]
Ops.. sorry, I forgot to add jars...   Fixed.

**Comment by kknishev** [comment]
@sevdokimov thanks a lot, but not working yet ((
```
logs = [
  {
    path: "**/*.log"  // All files with ".log" extension are visible by default
    format: {
      type: Log4jLogFormat
      locale: en_US // Set  custom locale (optional)
//      pattern: "%date{MMM [ ]d HH:mm:ss} %thread [%level] %msg%n"
      patterns: [
        "%date{MMM [ ]d HH:mm:ss} %thread %msg%n",
        "%date{MMM [ ]d HH:mm:ss} %thread %level %msg%n"
      ]
      customLevels: ["ALARM", "ATTANTION", "NORMAL"]
    }
  }
]
```
also trying with square brackets [%level]
check please with example.log

do I need to add colors here?
// Configure custom log levels. Where the group should be from 0 to 4.

**Comment by kknishev** [comment]
found!! this one is working now, thank you
```
      patterns: [
        "%date{MMM [ ]d HH:mm:ss} %thread [%level] %msg%n",
        "%date{MMM [ ]d HH:mm:ss} %thread %msg%n"
      ]
```

**Comment by kknishev** [comment]
another last question ))
is it possible to use two words for %level , like this: **REQUEST SUCCESS:**
`Sep 13 11:30:31 s9510-order-v2-5bc945cf7b-kznkv ws_orderv2[57]      ws_base.cpp:  62 REQUEST SUCCESS: register`

**Reply by sevdokimov** [reply]
It must work, you have to add `"REQUEST SUCCESS"` to `customLevels`, e.g. `customLevels: ["ALARM", "ATTANTION", "NORMAL", "REQUEST SUCCESS"]`

**Comment by kknishev** [comment]
hello! assist please how to use a word in quotes for %level, like this: "password", "userName"

trying this but no success
```
customLevels: ["\"password\""]
field-types.level.args."\"password\"".color = "#DB4437"
```

**Reply by sevdokimov** [reply]
Are "password", "userName" used as level field? Or you just want to highlight some words in the log meesage body?

**Reply by kknishev** [reply]
I want to highlight a word with quotes together, now I can highlight only word, without quotes
this words from the log meesage body

**Reply by sevdokimov** [reply]
You can add a regexp based highlighter to highlight any text:
```
ui-config {
  text-highlighters.myCustomHighlighterForUsername = {
     class: RegexHighlighter
     text-type: ["text"]
     args {
         regex = "\"userName\""
         style = {
             color: "#DB4437"
         }
     }
  }
}

```

**Reply by kknishev** [reply]
wow, thanks a lot, RegexHighlighter working well !!

**Reply by sevdokimov** [reply]
If the color is the same: `regex = "\"userName\"|\"password\"|\"another-word\""`
If the colors are different, you can create several `text-highlighters.$highlighterName = {...}` blocks

**Comment by kknishev** [comment]
Thank you @sevdokimov highlighet with regex working well.

Another little question:
`patterns` for http://localhost:8111/log?path=C:%5C/log_name.log working well (date, thread, levels.. )
but **the same** file from `log-paths` open from linux as http://localhost:8111/log?log=log_name show without patterns??

```
logs = [
  {
    path: "**/*.log"  // All files with ".log" extension are visible by default
    format: {
      type: Log4jLogFormat
      locale: en_US // Set  custom locale (optional)
      patterns: [
        "%date{MMM [ ]d HH:mm:ss} %thread %level: %msg%n",
        "%date{MMM [ ]d HH:mm:ss} %thread %level%n",
        "%date{MMM [ ]d HH:mm:ss} %thread %level %msg%n",
        "%date{MMM [ ]d HH:mm:ss} %thread [%level] %msg%n",
        "%date{MMM [ ]d HH:mm:ss} %thread %msg%n"
      ]
      customLevels: ["REQUEST SUCCESS", "REQUEST STARTED", "REQUEST FAILED", "REQUEST BEGIN"]
    }
  }

```

```
log-paths = {
  log_name = {
    file = "/var/log/test/log_name.log"
    host = ["10.1.10.10"] }
}
```

btw, a little mistake in config.conf exampels - extra log`s` in the path here: `/logs?log=` sould be `/log?log=`

//  // "my-app" is a query parameter in URL, http://localhost:8111/logs?log=my-app  shows logs from ~/my-app/logs/my-app.log
//  my-app = {
//    file = "/"
//  }
//  // "multinode-app" is a query parameter in URL, http://localhost:8111/logs?log=multinode-app shows logs from
//  // /opt/multinode-app/logs/app.log files on "node-cn-01", "node-cn-02" and "node-cn-03" hosts.

**Comment by kknishev** [comment]
@sevdokimov assist please
does logs `format:` available with `log-paths` ?
now log-paths shotcards shows as "log format: simple format: each line is an event"

**Reply by sevdokimov** [reply]
yes, `format:` is available with `log-paths`. Your `log-paths.log_name` has a reference to another host (10.1.10.10). Have you configured the log format on the remote machine (10.1.10.10)?

**Reply by kknishev** [reply]
No, `format:` configured on the local machine only, in the same config.conf where and  `log-paths`

On the host machile configured `log-viewer.backdoor_server.port` only, will it work in this case?

**Reply by sevdokimov** [reply]
You have to configure the list of visible logs and its format on the host machine. Each instance has it own `logs=[...]` section that contains information about which logs are visible on that machine and their format.

**Reply by kknishev** [reply]
yes, I see that `format:` is a part of `logs` section..
```
logs = [
    path: "**/*.log"
    format:
```
but is it possible to add `format:` in `log-paths` section also ?
this makes it possible for each user to use a personal log format

**Reply by sevdokimov** [reply, marked as answer]
It's possible to support `format:` in `log-paths` , you can create a separate issue.

**Comment by kknishev** [comment]
@sevdokimov could you please take a look ^^^

**Reply by sevdokimov** [reply]
Answered
